fe/fe-core/src/main/java/com/starrocks/http/rest/TransactionLoadAction.java

```diff
@@ -184,184 +213,189 @@ public void executeWithoutPassword(BaseRequest request, BaseResponse response) t
+            sendResult(request, response, result.getResult());
+            return;
         }
 
-        Long nodeID = null;
+        // redirect transaction op to BE
+        TNetworkAddress redirectAddress = result.getRedirectAddress();
+        if (null == redirectAddress) {
+            Long nodeId = getNodeId(txnOperation, label);
+            ComputeNode node = GlobalStateMgr.getCurrentState().getNodeMgr().getClusterInfo().getBackend(nodeId);
+            if (node == null) {
+                node = GlobalStateMgr.getCurrentState().getNodeMgr().getClusterInfo().getComputeNode(nodeId);
+                if (node == null) {
+                    throw new UserException("Node " + nodeId + " is not alive");
+                }
+            }
 
-        if (Strings.isNullOrEmpty(dbName)) {
-            throw new UserException("No database selected.");
-        }
-        if (Strings.isNullOrEmpty(label)) {
-            throw new UserException("empty label.");
+            redirectAddress = new TNetworkAddress(node.getHost(), node.getHttpPort());
         }
 
-        // 1. handle commit/rollback PREPARED transaction
-        if ((op.equalsIgnoreCase(TXN_COMMIT) || op.equalsIgnoreCase(TXN_ROLLBACK)) && channelIdStr == null) {
-            TransactionResult resp = new TransactionResult();
-            Database db = GlobalStateMgr.getCurrentState().getDb(dbName);
-            if (db == null) {
-                throw new UserException("database " + dbName + " not exists");
-            }
-            TransactionStatus txnStatus = GlobalStateMgr.getCurrentState().getGlobalTransactionMgr().getLabelStatus(db.getId(),
-                    label);
-            Long txnID = GlobalStateMgr.getCurrentState().getGlobalTransactionMgr().getLabelTxnID(db.getId(), label);
-            if (txnStatus == TransactionStatus.PREPARED) {
-                if (txnID == -1) {
-                    throw new UserException("label " + label + " txn not exist");
-                }
+        LOG.info("Redirect transaction action to destination={}, db: {}, table: {}, op: {}, label: {}",
+                redirectAddress, txnOperationParams.getDbName(), txnOperationParams.getTableName(), txnOperation, label);
+        redirectTo(request, response, redirectAddress);
+    }
 
-                if (op.equalsIgnoreCase(TXN_COMMIT)) {
-                    long timeoutMillis = 20000;
-                    if (timeout != null) {
-                        timeoutMillis = Long.parseLong(timeout) * 1000;
+    private TransactionOperationHandler getTxnOperationHandler(TransactionOperationParams params) {
+        if (params.getChannel().notNull()) {
+            return new TransactionWithChannelHandler(params);
+        }
+
+        LoadJobSourceType sourceType = params.getSourceType();
+        if (null == sourceType) {
+            // try to get source type from cache first
+            sourceType = txnSrcTypeCache.computeIfAbsent(params.getLabel(), label -> {
```

Comment:
Why not protect `txnSrcTypeCache` from the concurrent access? I think it's enough to only put labels of transaction stream load to the cache when begin, and doing it in `getNodeId` seems better?

Reply:
The `getNodeId` method will be called when redirect the request to BE, which is not applicable to the Bypass Write scenario.

Why i'm not protect`txnSrcTypeCache` from the concurrent access is that,  a cache miss results in the execution of the `getLabelTransactionState` method when access to `txnSrcTypeCache` concurrently is a boundary case, and even if it is accessed, it will not have much impact. However, most of the time the cache is hit, so locking access is inefficient.